hasktorch/src/Torch/Typed/Functional.hs

```diff
@@ -2267,14 +2270,125 @@ det
   -> Tensor device dtype (Det shape) -- ^ output
 det input = unsafePerformIO $ ATen.cast1 ATen.Managed.det_t input
 
--- diag_embed :: Tensor device dtype shape -> Int -> Int -> Int -> Tensor device dtype shape
--- diag_embed _input _offset _dim1 _dim2 = unsafePerformIO $ (ATen.cast4 ATen.Managed.diag_embed_tlll) _input _offset _dim1 _dim2
+type family DimsDistinctAscendingCheck (dim1 :: Nat) (dim2 :: Nat) (cmp :: Ordering) :: Constraint where
+  DimsDistinctAscendingCheck _ _ 'LT = ()
+  DimsDistinctAscendingCheck dim1 dim2 _ =
+    TypeError
+    ( Text "Dimensions must be distinct and in ascending order, but got "
+        :<>: ShowType dim1
+        :<>: Text ", "
+        :<>: ShowType dim2
+    )
+
+type family DimsDistinctAscending (dim1 :: Nat) (dim2 :: Nat) :: Constraint where
+  DimsDistinctAscending dim1 dim2 = DimsDistinctAscendingCheck dim1 dim2 (CmpNat dim1 dim2)
+
+type family DiagEmbedShapeImpl (dim1 :: Nat) (dim2 :: Nat) (shape :: [Nat]) (n :: Nat) :: [Nat] where
+  DiagEmbedShapeImpl dim1 dim2 shape n = Insert dim1 n (Insert (dim2 - 1) n (Init shape))
+
+type family DiagEmbedShape (index :: Nat) (dim1 :: Nat) (dim2 :: Nat) (shape :: [Nat]) :: [Nat] where
+  DiagEmbedShape index dim1 dim2 shape = DiagEmbedShapeImpl dim1 dim2 shape (Last shape + index)
+
+-- | diagEmbed
+--
+-- >>> dtype &&& shape $ diagEmbed @0 @0 @1 Upper (ones :: CPUTensor 'D.Float '[2,3])
+-- (Float,[2,3,3])
```

Comment:
![image](https://user-images.githubusercontent.com/2469618/87187104-95a54a80-c327-11ea-9622-94edde88f05a.png)
CI's error occurs here.
Should we change it to `(Float,[3,3,2])`?

Reply:
oops, let me check! I missed that this one started failing

Reply:
Ah, ok, I think the issue was that I incorrectly translated dimensions -2, -1 to 0, 1 (instead of the correct 1, 2) when I removed the dimension type I had introduced earlier. This should be fixed in 057a1d5.

Reply:
Thank you for your updating.
I am not familiar with negative dimention.
I think I now understand how to use it and why -2 becomes 0 ( `dim of input tensor` - 2 = 0).